README.md
jQuery Scanner Detection
========================

jQuery Scanner Detection is a small plugin to detect when user use a scanner (barcode, QR Code...) instead of a keyboard, and call specific callbacks.


How use it
----------
To initialize the detection, 2 ways:

    $(selector).scannerDetection(); // Initialize with default options
    $(selector).scannerDetection({...}); // Initialize with an object that contains options
    $(selector).scannerDetection(function(){...}); // Initialize with a function that is onComplete callback
    
To simulate a scanning after initialisation:

    $(selector).scannerDetection('scanned string');
	
To disable the detection (deinitialize):

    $(selector).scannerDetection(false);
	
See a detailed tutorial with presets for different scanner models and first steps to take when starting [here](http://a.kabachnik.info/jquery-scannerdetection-tutorial.html).


Options
-------

### onComplete 
Callback after detection of a successful scanning
### onError 
Callback after detection of a unsuccessful scanning (scanned string in parameter)
### onReceive 
Callback after receiving and processing a char (scanned char in parameter)
### onKeyDetect  
Callback after detecting a keyDown (key char in parameter) - in contrast to onReceive, this fires for non-character keys like tab, arrows, etc. too!
### onScanButtonLongPressed  
Callback after detection of a successfull scan while the scan button was pressed and held down. This can only be used if the scan button behaves as a key itself (see scanButtonKeyCode). This long press event can be used to add a secondary action. For example, if the primary action is to count some items with barcodes (e.g. products at goods-in), it is comes very handy if a number pad pops up on the screen when the scan button is held. Large number can then be easily typed it instead of scanning fifty times in a row. 
Note: this option requires scanButtonKeyCode to be set to a valid key code!
### timeBeforeScanTest
Default: 100  
Wait duration (ms) after keypress event to check if scanning is finished
### avgTimeByChar
Default: 30  
Average time (ms) between 2 chars. Used to do difference between keyboard typing and scanning
### minLength
Default: 6  
Minimum length for a scanning
### endChar
Default: [9,13]  
Chars to remove and means end of scanning
### startChar
Default: []  
Chars to remove and means start of scanning
### ignoreIfFocusOn
Default: false  
Ignore scans if the currently focused element matches this selector. Per example, if you set this option to 'input', scanner detection will be disable if an input is focused, when the scan occurs.
### scanButtonKeyCode
Default: null  
Key code of the scanner hardware button (if the scanner button a acts as a key itself). Knowing this key code is important, because it is not part of the scanned code and must be ignored. On the other hand, knowing it can be usefull: pressing the button multiple times fast normally results just in one scan, but you still could count the number of times pressed, allowing the user to input quantities this way (typical use case would be counting product at goods-in). 
### scanButtonLongPressThreshold
Default: 3  
You can let the user perform some special action by pressing and holding the scan button. In this case the button will issue multiple keydown events. This parameter sets, how many sequential events should be interpreted as holding the button down.  
### stopPropagation
Default: false  
Stop immediate propagation on keypress event
### preventDefault
Default: false  
Prevent default action on keypress event


Events
------
All callbacks are of type function and can also be bound as event listeners.  
Like this, you can add multiple callbacks for each event.  
Of course, events exist only if plugin is initialized on selector.

    $(selector)
        .bind('scannerDetectionComplete',function(e,data){...})
        .bind('scannerDetectionError',function(e,data){...})
        .bind('scannerDetectionReceive',function(e,data){...})

### scannerDetectionComplete
Callback after detection of a successful scanning  
Event data: {string: "scanned string"}
### scannerDetectionError
Callback after detection of a unsuccessful scanning  
Event data: {string: "scanned string"}  
### scannerDetectionReceive
Callback after receive and process a char  
Event data: {evt: {original keydown event}}
### scannerDetectionKeyDetect
Callback after detecting a keyDown - in contrast to onReceive, this fires for non-character keys like tab, arrows, etc. too!
Event data: {evt: {original keydown event}}

Requirements
------------
jQuery > 1.4.4. Note, that some odd side effects may occur before version 1.6, if your event handlers rely on using the event object (for example calling e.which from within the onKeyDetect handler).

Browser compatibility
---------------------
On old IE browser (IE<9), `Date.now()` and `Array.indexOf()` are not implemented.  
If you plan to use this plugin on these browsers, please add jquery.scannerdetection.compatibility.js file before the plugin.


CHANGELOG.txt
1.2.2
  The keydown event is now used for input characters instead of keypress
  Various internal code changes to use jQuery plugin best practices

1.2.1
  Add a onKeyDetect callback being called if any key (even non-chars) is detected
  Changed default scanButtonKeyCode to false (it turned out, that the previous value of '0' was actually used by some scanners)

1.2.0
  Add options startChar, ignoreIfFocusOn, scanButtonKeyCode, scanButtonLongPressThreshold

1.1.4
  Add script to extend compatibility to IE<9

1.1.3
  Change event.timestamp to Date.now() because Firefox has an issue with that

1.1.2
  Fix issue with tab key

1.1.1
  Fix issue with $(selector).scannerDetection('scanned string')

1.1.0
 Add events scannerDetectionComplete, scannerDetectionError, scannerDetectionReceive

1.0.0
 Initial version


jquery.scannerdetection.js
/*
 * jQuery Scanner Detection
 *
 * Copyright (c) 2013 Julien Maurel
 *
 * Licensed under the MIT license:
 * http://www.opensource.org/licenses/mit-license.php
 *
 * Project home:
 * https://github.com/julien-maurel/jQuery-Scanner-Detection
 *
 * Version: 1.2.2
 *
 */
(function($){
    $.fn.scannerDetection = function(options) {

        // If string given, call onComplete callback
        if(typeof options === "string") {
            return this.each(function(){
                this.scannerDetectionTest(options);
            });
        }

        // If false (boolean) given, deinitialize plugin
        if(options === false) {
            return this.each(function(){
                this.scannerDetectionOff();
            });
        }

        if(typeof options === "function"){
            options = {onComplete: options};
        }

        var settings = $.extend({
            onComplete: function() {}, // Callback after detection of a successful scanning (scanned string in parameter)
            onError: function() {}, // Callback after detection of a unsuccessful scanning (scanned string in parameter)
            onReceive: function() {}, // Callback after receiving and processing a char (scanned char in parameter)
            onKeyDetect: function() {}, // Callback after detecting a keyDown (key char in parameter) - in contrast to onReceive, this fires for non-character keys like tab, arrows, etc. too!
            timeBeforeScanTest: 100, // Wait duration (ms) after keydown event to check if scanning is finished
            avgTimeByChar: 30, // Average time (ms) between 2 chars. Used to do difference between keyboard typing and scanning
            minLength: 6, // Minimum length for a scanning
            endChar: [9,13], // Chars to remove and means end of scanning
            startChar: [], // Chars to remove and means start of scanning
            ignoreIfFocusOn: false, // do not handle scans if the currently focused element matches this selector
            scanButtonKeyCode: null, // Key code of the scanner hardware button (if the scanner button a acts as a key itself)
            scanButtonLongPressThreshold: 3, // How many times the hardware button should issue a pressed event before a barcode is read to detect a longpress
            onScanButtonLongPressed: function() {}, // Callback after detection of a successful scan while the scan button was pressed and held down
            stopPropagation: false, // Stop immediate propagation on keydown event
            preventDefault: false // Prevent default action on keydown event
        }, options || {});

        function getCharacterFromEvent(e) {
            var code = e.which;

            // These are special cases that don't fit the ASCII mapping
            var exceptions = {
                32: ' '.charCodeAt(0),
                186: 59, // ;
                187: 61, // =
                188: 44, // ,
                189: 45, // -
                190: 46, // .
                191: 47, // /
                192: 96, // `
                219: 91, // [
                220: 92, // \
                221: 93, // ]
                222: 39, // '
                //numeric keypad
                96: '0'.charCodeAt(0),
                97: '1'.charCodeAt(0),
                98: '2'.charCodeAt(0),
                99: '3'.charCodeAt(0),
                100: '4'.charCodeAt(0),
                101: '5'.charCodeAt(0),
                102: '6'.charCodeAt(0),
                103: '7'.charCodeAt(0),
                104: '8'.charCodeAt(0),
                105: '9'.charCodeAt(0),
                106: '*'.charCodeAt(0),
                107: '+'.charCodeAt(0),
                109: '-'.charCodeAt(0),
                110: '.'.charCodeAt(0),
                111: '/'.charCodeAt(0)
            };


            // Filter out non-alphanumeric key codes unless they're one of the exceptions
            if(code < 48 || code > 90) {
                if(exceptions[code] !== undefined) {
                    code = exceptions[code];
                } else {
                    return null
                }
            }

            var ch = String.fromCharCode(code);

            // If shifted translate characters, otherwise make lowercase
            if(e.shiftKey) {
                var special = {
                    1: '!',
                    2: '@',
                    3: '#',
                    4: '$',
                    5: '%',
                    6: '^',
                    7: '&',
                    8: '*',
                    9: '(',
                    0: ')',
                    ',': '<',
                    '.': '>',
                    '/': '?',
                    ';': ':',
                    "'": '"',
                    '[': '{',
                    ']': '}',
                    '\\': '|',
                    '`': '~',
                    '-': '_',
                    '=': '+'
                };

                if(special[ch] !== undefined) {
                    ch = special[ch];
                }
            } else {
                ch = ch.toLowerCase();
            }
            return ch;
        }

        return this.each(function() {
            var self = this,
                $self = $(self),
                firstCharTime = 0,
                lastCharTime = 0,
                stringWriting = '',
                callIsScanner = false,
                testTimer = false,
                scanButtonCounter = 0;

            self.initScannerDetection = function() {
                firstCharTime = 0;
                stringWriting = '';
                scanButtonCounter = 0;
            };

            self.isFocusOnIgnoredElement = function() {
                if (!settings.ignoreIfFocusOn) {
                    return false;
                }
                if(typeof settings.ignoreIfFocusOn === 'string') {
                    return $(':focus').is(options.ignoreIfFocusOn);
                }

                if(typeof settings.ignoreIfFocusOn === 'object' && settings.ignoreIfFocusOn.length > 0) {
                    var focused=$(':focus');
                    for(var i=0; i<settings.ignoreIfFocusOn.length; i++){
                        if(focused.is(settings.ignoreIfFocusOn[i])){
                            return true;
                        }
                    }
                }

                return false;
            };

            self.scannerDetectionOff = function() {
                $self.unbind('keydown.scannerDetection');
                $self.unbind('keypress.scannerDetection');
            };

            self.scannerDetectionTest = function(s) {
                // If string is given, test it
                if (s) {
                    firstCharTime = lastCharTime = 0;
                    stringWriting = s;
                }

                if (!scanButtonCounter) {
                    scanButtonCounter = 1;
                }

                var scanDuration = lastCharTime - firstCharTime;
                var scanDurationThreshold = stringWriting.length * settings.avgTimeByChar;

                // If the scan was long enough and done fast enough, it was a successful scan. Otherwise, error.
                if (stringWriting.length >= settings.minLength && scanDuration <= scanDurationThreshold) {
                    if (scanButtonCounter > settings.scanButtonLongPressThreshold) {
                        settings.onScanButtonLongPressed.call(self, stringWriting, scanButtonCounter);
                    } else {
                        settings.onComplete.call(self, stringWriting, scanButtonCounter);
                    }
                    $self.trigger('scannerDetectionComplete', {string: stringWriting});
                    self.initScannerDetection();
                    return true;
                } else {
                    settings.onError.call(self, stringWriting);
                    $self.trigger('scannerDetectionError', {string: stringWriting});
                    self.initScannerDetection();
                    return false;
                }
            };

            $self.unbind('.scannerDetection');

            $self.bind('keydown.scannerDetection', function(e) {

                if (this.isFocusOnIgnoredElement()) {
                    return;
                }

                // If it's just the button of the scanner, ignore it and wait for the real input
                if(settings.scanButtonKeyCode !== null && e.which === settings.scanButtonKeyCode) {
                    scanButtonCounter++;
                    e.preventDefault();
                    e.stopImmediatePropagation();

                    settings.onKeyDetect.call(self,e);
                    $self.trigger('scannerDetectionKeyDetect',{evt:e});
                    return;
                }



                if (settings.stopPropagation) {
                    e.stopImmediatePropagation();
                }

                if (settings.preventDefault) {
                    e.preventDefault();
                }

                // If it's not the first character and we encounter a terminating character, trigger scan process
                if (firstCharTime && settings.endChar.indexOf(e.which) !== -1) {
                    e.preventDefault();
                    e.stopImmediatePropagation();
                    callIsScanner = true;

                // If it's the first character and we encountered one of the starting characters, don't process the scan
                } else if(!firstCharTime && settings.startChar.indexOf(e.which) !== -1) {
                    e.preventDefault();
                    e.stopImmediatePropagation();
                    callIsScanner = false;

                // Otherwise, just add the character to the scan string we're building
                } else {
                    var character = getCharacterFromEvent(e);
                    if (character === null) {
                        return;
                    }
                    stringWriting += character;
                    callIsScanner = false;
                }

                if (!firstCharTime) {
                    firstCharTime = Date.now();
                }

                lastCharTime = Date.now();

                if (testTimer) {
                    clearTimeout(testTimer);
                }

                if (callIsScanner) {
                    self.scannerDetectionTest();
                    testTimer = false;
                } else {
                    testTimer = setTimeout(self.scannerDetectionTest,settings.timeBeforeScanTest);
                }

                settings.onReceive.call(self,e);
                $self.trigger('scannerDetectionReceive',{evt:e});
            });
        });
    }
})(jQuery);
